File: sdk/python/test/regression/run_service_test.py
import logging
import os
import unittest
import uuid

from kucoin_universal_sdk.api import DefaultClient
from kucoin_universal_sdk.generate.account.fee import GetBasicFeeReqBuilder, GetBasicFeeReq
from kucoin_universal_sdk.generate.earn.earn import GetSavingsProductsReqBuilder
from kucoin_universal_sdk.generate.futures.order import AddOrderReqBuilder, AddOrderReq, \
    GetOrderByOrderIdReqBuilder as FuturesGetOrderByOrderIdReqBuilder, CancelOrderByIdReqBuilder
from kucoin_universal_sdk.generate.margin.order import AddOrderReqBuilder as MarginAddOrderReqBuilder, \
    AddOrderReq as MarginAddOrderReq, \
    GetOrderByOrderIdReqBuilder as MarginGetOrderByOrderIdReqBuilder, \
    CancelOrderByOrderIdReqBuilder as MarginCancelOrderByOrderIdReqBuilder
from kucoin_universal_sdk.generate.spot.market import Get24hrStatsReqBuilder
from kucoin_universal_sdk.generate.spot.order import AddOrderSyncReqBuilder, AddOrderSyncReq, \
    GetOrderByOrderIdReqBuilder, CancelOrderByOrderIdReqBuilder
from kucoin_universal_sdk.model import ClientOptionBuilder, RestResponse
from kucoin_universal_sdk.model import GLOBAL_API_ENDPOINT, GLOBAL_FUTURES_API_ENDPOINT, \
    GLOBAL_BROKER_API_ENDPOINT
from kucoin_universal_sdk.model import TransportOptionBuilder


class CheckAllServiceTest(unittest.TestCase):

    def setUp(self):
        logging.basicConfig(
            level=logging.INFO,
            format='%(asctime)s %(levelname)s - %(message)s',
            datefmt='%Y-%m-%d %H:%M:%S'
        )

        #  Retrieve API secret information from environment variables
        key = os.getenv("API_KEY", "")
        secret = os.getenv("API_SECRET", "")
        passphrase = os.getenv("API_PASSPHRASE", "")

        # Set specific options, others will fall back to default values
        http_transport_option = (
            TransportOptionBuilder()
            .set_keep_alive(True)
            .set_max_pool_size(10)
            .set_max_connection_per_pool(10)
            .build()
        )

        # Create a client using the specified options
        client_option = (
            ClientOptionBuilder()
            .set_key(key)
            .set_secret(secret)
            .set_passphrase(passphrase)
            .set_spot_endpoint(GLOBAL_API_ENDPOINT)
            .set_futures_endpoint(GLOBAL_FUTURES_API_ENDPOINT)
            .set_broker_endpoint(GLOBAL_BROKER_API_ENDPOINT)
            .set_transport_option(http_transport_option)
            .build()
        )
        self.client = DefaultClient(client_option)
        self.rest_service = self.client.rest_service()

    def check_common_response(self, common_resp: RestResponse):
        self.assertEqual(common_resp.code, '200000')
        self.assertIsNotNone(common_resp.rate_limit)

    def test_account_service(self):
        account_service = self.rest_service.get_account_service()
        fee_api = account_service.get_fee_api()
        basic_fee_response = fee_api.get_basic_fee(
            GetBasicFeeReqBuilder().set_currency_type(GetBasicFeeReq.CurrencyTypeEnum.T_0).build())
        self.check_common_response(basic_fee_response.common_response)
        self.assertTrue(len(basic_fee_response.maker_fee_rate) > 0)
        self.assertTrue(len(basic_fee_response.taker_fee_rate) > 0)

    def test_earn_service(self):
        earn_service = self.rest_service.get_earn_service()
        earn_api = earn_service.get_earn_api()
        savings_products_response = earn_api.get_savings_products(
            GetSavingsProductsReqBuilder().set_currency("USDT").build())
        self.check_common_response(savings_products_response.common_response)
        self.assertTrue(len(savings_products_response.data) > 0)

    def test_margin_service(self):
        margin_service = self.rest_service.get_margin_service()
        order_api = margin_service.get_order_api()

        # add order
        add_order_req = MarginAddOrderReqBuilder().set_client_oid(uuid.uuid4().__str__()).set_side(
            MarginAddOrderReq.SideEnum.BUY).set_symbol(
            "BTC-USDT").set_type(MarginAddOrderReq.TypeEnum.LIMIT).set_price("10000").set_size("0.001").set_auto_repay(
            True).set_auto_borrow(True).set_is_isolated(True).build()
        add_order_resp = order_api.add_order(add_order_req)
        self.check_common_response(add_order_resp.common_response)
        self.assertTrue(len(add_order_resp.order_id) > 0)

        # query order
        query_order_req = MarginGetOrderByOrderIdReqBuilder().set_symbol("BTC-USDT").set_order_id(
            add_order_resp.order_id).build()
        query_order_resp = order_api.get_order_by_order_id(query_order_req)
        self.check_common_response(query_order_resp.common_response)
        self.assertTrue(len(query_order_resp.symbol), 0)

        # cancel order
        cancel_order_req = MarginCancelOrderByOrderIdReqBuilder().set_order_id(add_order_resp.order_id).set_symbol(
            "BTC-USDT").build()
        cancel_order_resp = order_api.cancel_order_by_order_id(cancel_order_req)
        self.check_common_response(cancel_order_resp.common_response)
        self.assertTrue(len(cancel_order_resp.order_id) > 0)

    def test_spot_service(self):
        spot_service = self.rest_service.get_spot_service()

        # query market api
        market_api = spot_service.get_market_api()
        stat_response = market_api.get24hr_stats(Get24hrStatsReqBuilder().set_symbol("BTC-USDT").build())
        self.check_common_response(stat_response.common_response)
        self.assertTrue(len(stat_response.last) > 0)

        order_api = spot_service.get_order_api()

        # add order
        add_order_sync_req = AddOrderSyncReqBuilder().set_client_oid(uuid.uuid4().__str__()).set_side(
            AddOrderSyncReq.SideEnum.BUY).set_symbol(
            "BTC-USDT").set_type(
            AddOrderSyncReq.TypeEnum.LIMIT).set_remark("sdk_test").set_price("10000").set_size("0.001").build()
        add_order_sync_resp = order_api.add_order_sync(add_order_sync_req)
        self.check_common_response(add_order_sync_resp.common_response)
        self.assertTrue(len(add_order_sync_resp.order_id) > 0)
        self.assertTrue(add_order_sync_resp.order_time > 0)

        # query order
        query_order_req = GetOrderByOrderIdReqBuilder().set_symbol("BTC-USDT").set_order_id(
            add_order_sync_resp.order_id).build()
        query_order_resp = order_api.get_order_by_order_id(query_order_req)
        self.check_common_response(query_order_resp.common_response)
        self.assertTrue(len(query_order_resp.symbol), 0)

        # cancel order
        cancel_order_req = CancelOrderByOrderIdReqBuilder().set_order_id(add_order_sync_resp.order_id).set_symbol(
            "BTC-USDT").build()
        cancel_order_resp = order_api.cancel_order_by_order_id(cancel_order_req)
        self.check_common_response(cancel_order_resp.common_response)
        self.assertTrue(len(cancel_order_resp.order_id) > 0)

    def test_futures_service(self):
        futures_service = self.rest_service.get_futures_service()

        # query market api
        market_api = futures_service.get_market_api()
        stat_response = market_api.get24hr_stats()
        self.check_common_response(stat_response.common_response)
        self.assertIsNotNone(stat_response.turnover_of24h)

        order_api = futures_service.get_order_api()

        # add order
        add_order_req = AddOrderReqBuilder().set_client_oid(uuid.uuid4().__str__()).set_side(
            AddOrderReq.SideEnum.BUY).set_symbol("XBTUSDTM").set_leverage(1).set_type(
            AddOrderReq.TypeEnum.LIMIT).set_remark("sdk_test").set_margin_mode(
            AddOrderReq.MarginModeEnum.CROSS).set_price("1").set_size(1).build()
        add_order_resp = order_api.add_order(add_order_req)
        self.check_common_response(add_order_resp.common_response)
        self.assertTrue(len(add_order_resp.order_id) > 0)

        # query order
        query_order_req = FuturesGetOrderByOrderIdReqBuilder().set_order_id(add_order_resp.order_id).build()
        query_order_resp = order_api.get_order_by_order_id(query_order_req)
        self.check_common_response(query_order_resp.common_response)
        self.assertTrue(len(query_order_resp.symbol), 0)

        # cancel order
        cancel_order_req = CancelOrderByIdReqBuilder().set_order_id(add_order_resp.order_id).build()
        cancel_order_resp = order_api.cancel_order_by_id(cancel_order_req)
        self.check_common_response(cancel_order_resp.common_response)
        self.assertTrue(len(cancel_order_resp.cancelled_order_ids) > 0)

File: sdk/node/example/src/regression/run_service_test.ts
import {randomUUID} from 'crypto';
import assert from 'assert';
import {
    Account,
    ClientOptionBuilder,
    DefaultClient,
    Earn,
    Futures,
    GlobalApiEndpoint,
    GlobalBrokerApiEndpoint,
    GlobalFuturesApiEndpoint,
    Margin,
    RestResponse,
    Spot,
    TransportOptionBuilder,
} from 'kucoin-universal-sdk';

const describe = (name: string, fn: () => Promise<void> | void) => {
    console.log(`\n${name}`);
    fn();
};
const it = (name: string, fn: () => Promise<void>) =>
    fn()
        .then(() => console.log(`  ✓ ${name}`))
        .catch((e) => {
            console.error(`  ✗ ${name}`);
            console.error(e);
            process.exitCode = 1;
        });

const httpOption = new TransportOptionBuilder()
    .setKeepAlive(true)
    .setMaxConnsPerHost(10)
    .build();

const client = new DefaultClient(
    new ClientOptionBuilder()
        .setKey(process.env.API_KEY ?? '')
        .setSecret(process.env.API_SECRET ?? '')
        .setPassphrase(process.env.API_PASSPHRASE ?? '')
        .setSpotEndpoint(GlobalApiEndpoint)
        .setFuturesEndpoint(GlobalFuturesApiEndpoint)
        .setBrokerEndpoint(GlobalBrokerApiEndpoint)
        .setTransportOption(httpOption)
        .build(),
);

const rest = client.restService();

function checkCommon(resp?: RestResponse) {
    assert.ok(resp !== undefined);
    if (resp) {
        assert.strictEqual(resp.code, '200000');
        assert.ok(resp.rateLimit);
    }
}

describe('CheckAllServiceTest', () => {
    //
    // Account
    //
    it('account_service', async () => {
        const feeApi = rest.getAccountService().getFeeApi();
        const resp = await feeApi.getBasicFee(
            Account.Fee.GetBasicFeeReq.builder()
                .setCurrencyType(Account.Fee.GetBasicFeeReq.CurrencyTypeEnum._0)
                .build(),
        );
        checkCommon(resp.commonResponse);
        assert.ok(resp.makerFeeRate.length > 0);
        assert.ok(resp.takerFeeRate.length > 0);
    });

    //
    // Earn
    //
    it('earn_service', async () => {
        const earnApi = rest.getEarnService().getEarnApi();
        const resp = await earnApi.getSavingsProducts(
            Earn.Earn.GetSavingsProductsReq.builder().setCurrency('USDT').build(),
        );
        checkCommon(resp.commonResponse);
        assert.ok(resp.data.length > 0);
    });

    //
    // Margin
    //
    it('margin_service', async () => {
        const orderApi = rest.getMarginService().getOrderApi();

        const addReq = Margin.Order.AddOrderReq.builder()
            .setClientOid(randomUUID())
            .setSide(Margin.Order.AddOrderReq.SideEnum.BUY)
            .setSymbol('BTC-USDT')
            .setType(Margin.Order.AddOrderReq.TypeEnum.LIMIT)
            .setPrice('10000')
            .setSize('0.001')
            .setAutoRepay(true)
            .setAutoBorrow(true)
            .setIsIsolated(true)
            .build();
        const addResp = await orderApi.addOrder(addReq);
        checkCommon(addResp.commonResponse);
        assert.ok(addResp.orderId.length > 0);

        const queryReq = Margin.Order.GetOrderByOrderIdReq.builder()
            .setSymbol('BTC-USDT')
            .setOrderId(addResp.orderId)
            .build();
        const queryResp = await orderApi.getOrderByOrderId(queryReq);
        checkCommon(queryResp.commonResponse);
        assert.ok(queryResp.symbol.length > 0);

        const cancelReq = Margin.Order.CancelOrderByOrderIdReq.builder()
            .setOrderId(addResp.orderId)
            .setSymbol('BTC-USDT')
            .build();
        const cancelResp = await orderApi.cancelOrderByOrderId(cancelReq);
        checkCommon(cancelResp.commonResponse);
        assert.ok(cancelResp.orderId.length > 0);
    });

    //
    // Spot
    //
    it('spot_service', async () => {
        const spotService = rest.getSpotService();

        const statResp = await spotService
            .getMarketApi()
            .get24hrStats(
                Spot.Market.Get24hrStatsReq.builder().setSymbol('BTC-USDT').build(),
            );
        checkCommon(statResp.commonResponse);
        assert.ok(statResp.last.length > 0);

        const orderApi = spotService.getOrderApi();

        const addReq = Spot.Order.AddOrderSyncReq.builder()
            .setClientOid(randomUUID())
            .setSide(Spot.Order.AddOrderSyncReq.SideEnum.BUY)
            .setSymbol('BTC-USDT')
            .setType(Spot.Order.AddOrderSyncReq.TypeEnum.LIMIT)
            .setRemark('sdk_test')
            .setPrice('10000')
            .setSize('0.001')
            .build();
        const addResp = await orderApi.addOrderSync(addReq);
        checkCommon(addResp.commonResponse);
        assert.ok(addResp.orderId.length > 0);
        assert.ok(addResp.orderTime > 0);

        const queryReq = Spot.Order.GetOrderByOrderIdReq.builder()
            .setSymbol('BTC-USDT')
            .setOrderId(addResp.orderId)
            .build();
        const queryResp = await orderApi.getOrderByOrderId(queryReq);
        checkCommon(queryResp.commonResponse);
        assert.ok(queryResp.symbol.length > 0);

        const cancelReq = Spot.Order.CancelOrderByOrderIdReq.builder()
            .setOrderId(addResp.orderId)
            .setSymbol('BTC-USDT')
            .build();
        const cancelResp = await orderApi.cancelOrderByOrderId(cancelReq);
        checkCommon(cancelResp.commonResponse);
        assert.ok(cancelResp.orderId.length > 0);
    });

    //
    // Futures
    //
    it('futures_service', async () => {
        const futuresService = rest.getFuturesService();

        const statResp = await futuresService.getMarketApi().get24hrStats();
        checkCommon(statResp.commonResponse);
        assert.ok(statResp.turnoverOf24h);

        const orderApi = futuresService.getOrderApi();

        const addReq = Futures.Order.AddOrderReq.builder()
            .setClientOid(randomUUID())
            .setSide(Futures.Order.AddOrderReq.SideEnum.BUY)
            .setSymbol('XBTUSDTM')
            .setLeverage(1)
            .setType(Futures.Order.AddOrderReq.TypeEnum.LIMIT)
            .setRemark('sdk_test')
            .setMarginMode(Futures.Order.AddOrderReq.MarginModeEnum.CROSS)
            .setPrice('1')
            .setSize(1)
            .build();
        const addResp = await orderApi.addOrder(addReq);
        checkCommon(addResp.commonResponse);
        assert.ok(addResp.orderId.length > 0);

        const queryReq = Futures.Order.GetOrderByOrderIdReq.builder()
            .setOrderId(addResp.orderId)
            .build();
        const queryResp = await orderApi.getOrderByOrderId(queryReq);
        checkCommon(queryResp.commonResponse);
        assert.ok(queryResp.symbol.length > 0);

        const cancelReq = Futures.Order.CancelOrderByIdReq.builder()
            .setOrderId(addResp.orderId)
            .build();
        const cancelResp = await orderApi.cancelOrderById(cancelReq);
        checkCommon(cancelResp.commonResponse);
        assert.ok(cancelResp.cancelledOrderIds.length > 0);
    });
});
